composer/algorithms/algorithm_hparams.py

```diff
@@ -1,64 +1,90 @@
 # Copyright 2022 MosaicML Composer authors
 # SPDX-License-Identifier: Apache-2.0
 
-from __future__ import annotations
-
 import os
-from abc import ABC, abstractmethod
-from dataclasses import dataclass
-from typing import Optional
+from typing import Dict, List, Optional, Type, Union
 
 import yahp as hp
 
 import composer
+from composer.algorithms.agc import AGC
+from composer.algorithms.alibi import Alibi
+from composer.algorithms.augmix import AugMix
+from composer.algorithms.blurpool import BlurPool
+from composer.algorithms.channels_last import ChannelsLast
+from composer.algorithms.colout import ColOut
+from composer.algorithms.cutmix import CutMix
+from composer.algorithms.cutout import CutOut
+from composer.algorithms.ema import EMA
+from composer.algorithms.factorize import Factorize
+from composer.algorithms.ghost_batchnorm import GhostBatchNorm
+from composer.algorithms.label_smoothing import LabelSmoothing
+from composer.algorithms.layer_freezing import LayerFreezing
+from composer.algorithms.mixup import MixUp
+from composer.algorithms.no_op_model import NoOpModel
+from composer.algorithms.progressive_resizing import ProgressiveResizing
+from composer.algorithms.randaugment import RandAugment
+from composer.algorithms.sam import SAM
+from composer.algorithms.selective_backprop import SelectiveBackprop
+from composer.algorithms.seq_length_warmup import SeqLengthWarmup
+from composer.algorithms.squeeze_excite import SqueezeExcite
+from composer.algorithms.stochastic_depth import StochasticDepth
+from composer.algorithms.swa import SWA
 from composer.core.algorithm import Algorithm
 
+algorithm_registry: Dict[str, Union[Type[Algorithm], Type[hp.AutoInitializedHparams]]] = {
```

Comment:
I don't have a strong opinion here, but might be worth a quick design discussion on whether to keep the registry centralized (like we always have) vs. having a a decorator for each class, e.g. something like:

```
@hp.auto_register("blurpool", "algorithm")
class BlurPool:
```

If we keep this "menu" of options, its easy reference, but users would (1) need to know where the codepath of this registry lives in composer to find out this menu of options, and (2) to extend it, know where the specific registry lives and import to modify. Both of these are not very discoverable pieces of information at the moment. 

The decorator approach (1) keeps the key information with the class itself, which is more intuitive than some arbitrary registry file, and (2) keeps the code independant (e.g. a new contributed algorithm would register itself, without having to import/modify the registry code), and (3) is how external users would also register their own stuff. However, you do lose the centralized menu, and also litters our code with decorators. 

cc: @dblalock @eracah @moinnadeem @A-Jacobson who probably have some opinions here.

Reply:
I would prefer a registry. Though hanlin said all that needs to be said about it. The strongest point (which he already made) to me is that it aligns internal and external usage patterns. 

Losing the dictionary doesn't seem  like a large downside since external users would have to know it's there to refer to it. To learn about our algorithms they're likely looking at docs first then code.

Reply:
The decorator approach won't work because of how python imports work. Using some sort of autoregistration mechanism would result in phantom imports or offload more work on the entrypoint script. 

Consider the following:

```python
# File `wandb_logger.py`
@hp.register("logger")
class WandBLogger:
    ....
```

```python
# File `my_entrypoint.py`
import yahp as hp

from composer.trainer.trainer_hparams import TrainerHparams  # does NOT import wandb_logger.py

TrainerHparams.create({"loggers": {"wandb": ...}})
```

The above code would not work, as `TrainerHparams` would not know about the `WandBLogger` class.

Now, consider:
```python
# File `my_entrypoint.py`
import yahp as hp
import composer.loggers.wandb_logger  # phantom import

from composer.trainer.trainer_hparams import TrainerHparams

TrainerHparams.create({"loggers": {"wandb": ...}})
```
This would work, because `wandb_logger.py` was imported, calling the call to `hp.register`.

Now, consider this:
```python
# File `my_entrypoint.py`
import yahp as hp

from composer.trainer.trainer_hparams import TrainerHparams

hp.register_subclasses("composer.loggers.wandb_logger")  # explicitly tell yahp to go import this module to discover implementations

TrainerHparams.create({"loggers": {"wandb": ...}})
```
This would also work -- it's basically the same as the above, but offloads the phantom import into yahp. However, I'm also not a fan of this approach, as now every entrypoint would need to add a call to `hp.register_subclass`, and having the decorator + `hp.register_subclass` call is strictly more code than inserting something into a dictionary.

Going to resolve this conversation since I think this needs more design -- but open to alternatives that do not result in phantom imports and that do not involve more code than the current dictionary approach.

Reply:
What if TrainerHparams does a couple phantom imports: 
```from composer import algorithms
     from composer import loggers
     from composer import callbacks```, which because of the `__init__.py` file in each module will import every Algorithm, Logger, and Callback, which will in turn trigger all the import-time decoration.

Reply:
Actually, all it needs to do is: ```import composer```, which will trigger everything. Wdyt @A-Jacobson, @ravi-mosaicml?